parser/prism/Translator.cc

```diff
@@ -956,7 +955,30 @@ std::unique_ptr<parser::Hash> Translator::translateHash(pm_node_t *node, pm_node
                                                         bool isUsedForKeywordArguments) {
     pm_location_t *loc = &node->location;
 
-    auto sorbetElements = translateMulti(elements);
+    auto prismElements = absl::MakeSpan(elements.nodes, elements.size);
+
+    parser::NodeVec sorbetElements{};
+    sorbetElements.reserve(prismElements.size());
+
+    for (auto &pair : prismElements) {
+        if (PM_NODE_TYPE_P(pair, PM_ASSOC_SPLAT_NODE)) {
+            auto prismSplatNode = reinterpret_cast<pm_assoc_splat_node *>(pair);
+            pm_location_t *loc = &prismSplatNode->base.location;
+            auto value = translate(prismSplatNode->value);
+
+            if (value == nullptr) { // An anonymous splat like `f(**)`
+                auto sorbetSplatNode = make_unique<parser::ForwardedKwrestArg>(parser.translateLocation(loc));
+                sorbetElements.emplace_back(std::move(sorbetSplatNode));
+            } else { // Splatting an expression like `f(**h)`
+                auto sorbetSplatNode = make_unique<parser::Kwsplat>(parser.translateLocation(loc), std::move(value));
+                sorbetElements.emplace_back(std::move(sorbetSplatNode));
+            }
```

Comment:
Ideally, I would have wanted something more like this:

```suggestion
            std::unique_ptr<parser::Node> sorbetSplatNode;
            
            if (value == nullptr) { // An anonymous splat like `f(**)`
               sorbetSplatNode = make_unique<parser::ForwardedKwrestArg>(parser.translateLocation(loc));
            } else { // Splatting an expression like `f(**h)`
               sorbetSplatNode = make_unique<parser::Kwsplat>(parser.translateLocation(loc), std::move(value));
            }
            
            sorbetElements.emplace_back(std::move(sorbetSplatNode));
```

But remarkably, there's no easy way to cast a `std::unique_ptr<parser::ForwardedKwrestArg>` to `std::unique_ptr<parser::Node>`, like you can with "regular" C pointers. You'd have add your own helper, like this [`static_unique_ptr_cast()`](https://stackoverflow.com/a/33649683).

Reply:
That's so weird. I feel like we do something similar in the `translateAssignment` template, but maybe there's some subtlety there that makes it different? https://github.com/Shopify/sorbet/blob/prism/parser/prism/Translator.cc#L30

Reply:
Huh, I tried this again, exactly as my suggestion said, and it worked perfectly. 🤔